### DISCUSSION TITLE
GetServerSideProps + Router + CDN partial cache without query params

### DISCUSSION THREAD

**Discussion by Metroninja** [vercel/next.js#17646] [Help]
Scenario (we are currently facing in production) - 
You have a page where most of the data can be SSR, but there is session specific data in the query string you can't cache  - say a google gclid/msclid or any ad specific params/ or user specific data in the query string.  To save on performance cost and speed you place a cloud CDN in front of it (say Google Cloud CDN).

Ex: https://yoursite/location/seattle?gclid=abc123
lets say the page structure of this is `page/location/[location]`
you can set your cache to ignore all query params, but when you SSR the page with nextJS one of two things will happen depending on the proxy type -

1. The proxy will strip all query args and next/router will come into getServerSideProps with the query param `location` and nothing else.  This would be fine, except that  after the page hydrates (and forever after) your gclid is lost and can't be retrieved from next/router.  It will be there in `router.asPath` but there is no way to tell the router to stop using the `NEXT_DATA` provided value and to re-evaluate based on the url param (as far as I have been able to find).

2. next/router will get the query params as forward by the CDN and getServerSideProps will receive both the location and gclid query params.  This is also fine, as the gclid shouldn't do anything during SSR, and when the page finishes rendering the CDN caches the returned page with the glid in NEXT_DATA.  now user 2 comes along 30 seconds later with a new gclid, gets the CDN returned result that has the previous users gclid in NEXT_DATA and next/router will forever return the old gclid.  The router.asPath will have the new, proper gclid but there is no way to tell next/router to flush/recheck/compare it's query params with that of asPath.

Now there are obviously solutions here, but none are great.
1.  turn off CDN cacheing for this page (what we are currently doing).
2.  don't use next/router on this page, just use the window.location.search and parse it yourself.  not ideal.
3.  have a useEffect on router.asPath and parse the query yourself, maintaining query params in state.  Doable but not ideal as this feels like something that should be built in.
4.  on mount check for existence of a 'flush' url param that you set if not present to force nextJs to update the router client side.

Solutions I would hope for would be calling `useRouter()` with a param like `useRouter({revalidateOnMount: true})` and it would perform something like #3 above.

Is there something I'm missing, or has anyone encountered a similar situation?  I can't imagine this problem hasn't been encountered numerous times already in the past?

**Comment by mkarusell** [comment]
I'm having the exact same issue and without realizing it, have been running scenario 2 for quite some time, resulting in a bunch of issues. It took a long time to figure out what was wrong.

:(

**Comment by benkingcode** [comment]
This is a big issue for us with Stripe payment intent redirects, where it adds payment intent information to the query params. Any news on a fix for this?
